vowpalwabbit/vw.h

```diff
@@ -135,41 +135,26 @@ void save_predictor(vw& all, io_buf& buf);
 // First create the hash of a namespace.
 inline uint64_t hash_space(vw& all, const std::string& s)
 {
-  substring ss;
-  ss.begin = (char*)s.c_str();
-  ss.end = ss.begin + s.length();
-  return all.p->hasher(ss, all.hash_seed);
+  return all.p->hasher(s.data(), s.length(), all.hash_seed);
 }
 inline uint64_t hash_space_static(const std::string& s, const std::string& hash)
 {
-  substring ss;
-  ss.begin = (char*)s.c_str();
-  ss.end = ss.begin + s.length();
-  return getHasher(hash)(ss, 0);
+  return getHasher(hash)(s.data(), s.length(), 0);
 }
 // Then use it as the seed for hashing features.
 inline uint64_t hash_feature(vw& all, const std::string& s, uint64_t u)
 {
-  substring ss;
-  ss.begin = (char*)s.c_str();
-  ss.end = ss.begin + s.length();
-  return all.p->hasher(ss, u) & all.parse_mask;
+  return all.p->hasher(s.data(), s.length(), u) & all.parse_mask;
 }
 inline uint64_t hash_feature_static(const std::string& s, uint64_t u, const std::string& h, uint32_t num_bits)
 {
-  substring ss;
-  ss.begin = (char*)s.c_str();
-  ss.end = ss.begin + s.length();
   size_t parse_mark = (1 << num_bits) - 1;
-  return getHasher(h)(ss, u) & parse_mark;
+  return getHasher(h)(s.data(), s.length(), u) & parse_mark;
 }
 
 inline uint64_t hash_feature_cstr(vw& all, char* fstr, uint64_t u)
 {
-  substring ss;
-  ss.begin = fstr;
-  ss.end = ss.begin + strlen(fstr);
-  return all.p->hasher(ss, u) & all.parse_mask;
+  return all.p->hasher(fstr, strlen(fstr), u) & all.parse_mask;
```

Comment:
I approve of all the simplification here :-)

Reply:
This was originally even simpler but slim couldn't handle string_views, so I had to convert it to a more generic interface